### ISSUE TITLE
When jump to definition, it will jump into a file that is not in the compile_commands.json database, expecting to jump into AAA but actually jumping into BBB. It only happens if BBB is opened on once

### ISSUE THREAD

**Issue by sunqianGitHub** [clangd/clangd#1516]
![图片](https://user-images.githubusercontent.com/30254405/221108712-46b00196-2628-4ebb-b4d2-d806b564d35e.png)

early_chip_init() function is defined in both files with the same name chip_customization.c, and  two files Store in a different directory **aaa** and **bbb**

When click the function early_chip_init()  in bootloader.c to jump to definition, it will jump into a file that is not in the jason database, expecting to jump into AAA but actually jumping into BBB. It only happens if BBB is opened on once ; The Issue is the same whether BBB is opened on before or after AAA

If never opened BBB, can jump to AAA correctly

**Logs**

 ASTWorker building file /boot/bootloader/customization/**aaa**/chip_customization.c version 1 with command inferred from customization/**bbb**/chip_customization.c
[/........./boot/bootloader]
**System information**
ubuntu 20.04 server and windows Visual Studio Code version 1.75
Output of `clangd --version`:
.vscode-server/data/User/globalStorage/llvm-vs-code-extensions.vscode-clangd/install/15.0.6/clangd_15.0.6/lib/clang/15.0.6
Editor/LSP plugin:
clangd
v0.1.23
LLVM
C/C++ completion, navigation, and insights
Operating system: windows 10

**Comment by HighCommander4** [comment]
I guess this is a consequence of https://github.com/clangd/clangd/issues/1258

**Comment by MNASTM** [comment]
Ok this is related to [#1258](https://github.com/clangd/clangd/issues/1258) and I fully understand the explanations from @sam-mccall , but here we are not really talking about multi projects yet and I am wondering if there can be a way out?

In my use case I explicitly give the compile_commands.json path to the server (--compile-commands-dir option) so I was expecting no heuristic, especially not from files outside its scope. But still, I have the exact same behavior as describe above. 

First I just wanted to confirm this is still the expected behavior? 
If yes, don't you think this can be improved in this case? (either on server or vscode extension side?)

**Comment by HighCommander4** [comment]
> First I just wanted to confirm this is still the expected behavior? If yes, don't you think this can be improved in this case? (either on server or vscode extension side?)

It would help to have a concrete example that can be played around with locally, to reason about and try to answer these questions.

**Comment by MNASTM** [comment]
Sure, here is a basic [repo](https://github.com/MNASTM/clangd-issue-682) to give a try

After few other trials on my side, it sounds like the issue is slightly different from what was originally stated:

"It only happens if BBB is opened on once; The issue is the same whether BBB is opened on before or after AAA"

Instead, I have: 

"It only happens if BBB is opened once and AAA is never opened; No issue if both have been opened, no matter the order."

Here are some clangd logs when this happens:

```
...
When the server starts and pack-div@1.0.0/div.c is open
...
I[13:39:30.501] clangd version 18.1.3 (https://github.com/llvm/llvm-project c13b7485b87909fcf739f62cfa382b55407433c0)
I[13:39:30.502] Features: windows+grpc
I[13:39:30.502] PID: 32188
I[13:39:30.502] Working directory: c:\Dev\MNASTM\clangd-issue-682
I[13:39:30.502] argv[0]: c:\Users\naudmatt\AppData\Roaming\Code\User\globalStorage\llvm-vs-code-extensions.vscode-clangd\install\18.1.3\clangd_18.1.3\bin\clangd.exe
I[13:39:30.502] argv[1]: --compile-commands-dir
I[13:39:30.502] argv[2]: C:\Dev\MNASTM\clangd-issue-682\build
I[13:39:30.515] Starting LSP over stdin/stdout
I[13:39:30.565] <-- initialize(0)
I[13:39:30.570] --> reply:initialize(0) 5 ms
I[13:39:30.599] <-- initialized
I[13:39:30.614] <-- textDocument/didOpen
I[13:39:30.617] Loaded compilation database from C:\Dev\MNASTM\clangd-issue-682\build\compile_commands.json
I[13:39:30.618] <-- textDocument/documentSymbol(1)
I[13:39:30.618] ASTWorker building file c:\Dev\MNASTM\clangd-issue-682\pack-div@1.0.0\src\div.c version 1 with command inferred from C:\Dev\MNASTM\clangd-issue-682\pack-div@1.1.0\src\div.c
[C:/Dev/MNASTM/clangd-issue-682/build]
"C:\\clang+llvm-18.1.6-x86_64-pc-windows-msvc\\bin\\clang.exe" -IC:/Dev/MNASTM/clangd-issue-682/pack-div@1.1.0/inc -IC:/Dev/MNASTM/clangd-issue-682/./inc -O0 -g -Xclang -gcodeview -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -c "-resource-dir=C:\\Users\\naudmatt\\AppData\\Roaming\\Code\\User\\globalStorage\\llvm-vs-code-extensions.vscode-clangd\\install\\18.1.3\\clangd_18.1.3\\lib\\clang\\18" -- "c:\\Dev\\MNASTM\\clangd-issue-682\\pack-div@1.0.0\\src\\div.c"
I[13:39:30.644] --> window/workDoneProgress/create(0)
I[13:39:30.644] Enqueueing 2 commands for indexing
I[13:39:30.645] <-- textDocument/codeAction(2)
I[13:39:30.646] <-- textDocument/documentLink(3)
I[13:39:30.646] <-- textDocument/inlayHint(4)
I[13:39:30.647] <-- reply(0)
I[13:39:30.647] --> $/progress
I[13:39:30.647] --> $/progress
I[13:39:30.649] --> $/progress
I[13:39:30.649] --> $/progress
I[13:39:30.668] --> textDocument/clangd.fileStatus
I[13:39:30.693] Built preamble of size 231600 for file c:\Dev\MNASTM\clangd-issue-682\pack-div@1.0.0\src\div.c version 1 in 0.02 seconds
I[13:39:30.694] --> workspace/semanticTokens/refresh(1)
I[13:39:30.696] <-- reply(1)
I[13:39:30.702] Indexing c17 standard library in the context of c:\Dev\MNASTM\clangd-issue-682\pack-div@1.0.0\src\div.c
I[13:39:30.706] <-- textDocument/semanticTokens/full(5)
....
User intereaction => open main.c
....
I[13:46:12.718] <-- textDocument/didOpen
I[13:46:12.720] ASTWorker building file c:\Dev\MNASTM\clangd-issue-682\src\main.c version 1 with command 
[C:/Dev/MNASTM/clangd-issue-682/build]
"C:\\clang+llvm-18.1.6-x86_64-pc-windows-msvc\\bin\\clang.exe" -IC:/Dev/MNASTM/clangd-issue-682/pack-div@1.1.0/inc -IC:/Dev/MNASTM/clangd-issue-682/./inc -O0 -g -Xclang -gcodeview -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -o "CMakeFiles\\app1-sw1.dir\\src\\main.c.obj" -c "-resource-dir=C:\\Users\\naudmatt\\AppData\\Roaming\\Code\\User\\globalStorage\\llvm-vs-code-extensions.vscode-clangd\\install\\18.1.3\\clangd_18.1.3\\lib\\clang\\18" -- "c:\\Dev\\MNASTM\\clangd-issue-682\\src\\main.c"
I[13:46:12.722] <-- textDocument/codeAction(21)
I[13:46:12.724] <-- textDocument/documentSymbol(22)
I[13:46:12.728] --> textDocument/clangd.fileStatus
I[13:46:12.748] Built preamble of size 231800 for file c:\Dev\MNASTM\clangd-issue-682\src\main.c version 1 in 0.02 seconds
I[13:46:12.748] --> workspace/semanticTokens/refresh(2)
I[13:46:12.749] <-- reply(2)
....
User interaction => Go to definition on div(6,3) => leads to opening the wrong file (pack-div@1.0.0\src\div.c)
....
I[13:47:50.606] <-- textDocument/definition(38)
I[13:47:50.607] --> reply:textDocument/definition(38) 0 ms
I[13:47:50.607] --> textDocument/clangd.fileStatus
I[13:47:50.623] <-- textDocument/documentHighlight(39)
I[13:47:50.623] --> reply:textDocument/documentHighlight(39) 0 ms
I[13:47:50.623] --> textDocument/clangd.fileStatus
I[13:47:50.623] <-- textDocument/codeAction(40)
I[13:47:50.624] --> reply:textDocument/codeAction(40) 0 ms
I[13:47:50.624] --> textDocument/clangd.fileStatus
I[13:47:50.648] <-- textDocument/documentLink(41)
I[13:47:50.648] <-- textDocument/inlayHint(42)
I[13:47:50.648] --> reply:textDocument/documentLink(41) 0 ms
I[13:47:50.649] --> reply:textDocument/inlayHint(42) 0 ms
I[13:47:50.649] --> textDocument/clangd.fileStatus
I[13:47:50.829] <-- textDocument/foldingRange(43)
I[13:47:50.829] --> reply:textDocument/foldingRange(43) 0 ms
I[13:47:50.876] <-- textDocument/codeAction(44)
I[13:47:50.877] --> reply:textDocument/codeAction(44) 0 ms
I[13:47:50.877] --> textDocument/clangd.fileStatus
I[13:47:50.928] <-- textDocument/semanticTokens/full/delta(45)
I[13:47:50.928] <-- textDocument/documentSymbol(46)
I[13:47:50.928] --> reply:textDocument/semanticTokens/full/delta(45) 0 ms
I[13:47:50.928] --> reply:textDocument/documentSymbol(46) 0 ms
I[13:47:50.928] --> textDocument/clangd.fileStatus
```

I am not familiar at all with the server implementation but to me there is a heuristic issue as long as the server has the compile_commands.json information from the beginning which should allow to link the main.c to the right div.c, no matter what other files have been opened, or no matter if the right file has not yet been opened, WDYT?

**Comment by HighCommander4** [comment]
> Sure, here is a basic [repo](https://github.com/MNASTM/clangd-issue-682) to give a try

Thanks, this is very helpful.

I can confirm that I see the same behaviour that you describe in the repo's README.

It's not clear to me off the top of my head what is the mechanism that gives rise to this behaviour. It will require some debugging to understand that. I will do that as time permits and follow up here.

**Comment by HighCommander4** [comment]
Ok, I've debugged this a bit, and I believe I now understand how the observed behaviour arises.

As some background, clangd's symbol index has two parts:
 1. the **dynamic index**, which is in-memory only, and contains symbols from files opened since clangd was started
 2. the **background index**, which is persisted to disk, and contains symbols from all files that appear in `compile_commands.json`

When looking up a symbol in the index, it's looked up in both the dynamic index and the background index. If it's found in both places, the results are merged. During a merge, if both input symbols have a definition location (one of the properties of a symbol), the definition location from the dynamic index is retained in the resulting merged symbol.

So let's consider the scenarios from the example project README:

* Scenario 1 is the simplest. The dynamic index does contain any `div` symbol with a definition. The background index contains the `div` symbol from pack-div@1.1.0 (since that is the file that was present in the `compile_commands.json` and thus added to the background index), so the definition in pack-div@1.1.0 is returned.
* Scenario 2 is the same as Scenario 1; the fact that pack-div@1.0.0/div.c was opened during a previous clangd session does not matter, since the dynamic index does not persist between clangd sessions.
* Next, let's look at the last ("NOK") scenario. The dynamic index contains the `div` symbol from pack-div@1.0.0. The background index contains the `div` symbol from pack-div@1.1.0. When the two symbols are merged, the definition location from the dynamic index, pack-div@1.0.0, is retained, so that's what we get.
* Finally, what happens in scenarios 3 and 4? Here, the copy of the symbol in the dynamic index is itself the result of merging the two instances of the symbol that come from the two opened files. For this merge, it's a matter of luck which location wins. It ends up boiling down to the order of iterating a hashtable where the keys are the full paths of the files. It so happens that for these two filenames, pack-div@1.1.0 wins over pack-div@1.0.0. But if I change pack-div@1.0.0 to pack-div@1.2.0, now pack-div@1.2.0 wins consistently, and we get the "undesirable" result regardless of what order the files were opened.

**Comment by HighCommander4** [comment]
So, to circle back to the original questions:

> this is still the expected behavior? If yes, don't you think this can be improved in this case?

The choice to prefer the definition from the dynamic index when merging symbols does seem to be deliberate.

However, I expect it's motivated by a different scenario: one where the dynamic index contains symbols from parsing (a subset of) the same files as the background index, but the dynamic index is **newer** (e.g. it may reflect file changes that have not been reflected in the background index yet).

Information in the dynamic index conflicting with information in the background index because it's coming from two different files that define symbols with the same name is, I believe, not a scenario that has been designed for at all (as discussed in #1258).

So, I think there may potentially be room for enhancing the symbol merging logic with heuristics related to preferring symbols coming from files in `compile_commands.json`. It will require some thought in figuring out what exactly the heuristic should be, making sure it doesn't break existing dynamic index use cases, and that it can be implemented efficiently (merging symbols seems like a hot codepath).

**Comment by VGRSTM** [comment]
@HighCommander4 thank you so much for such analysis !

So first if my understanding is right we can claim current behavior is related to `clangd` server, not `vscode-clangd` client: do you confirm ?

Considering, yes, some "undesirable" result do you foresee a way to get rid of with current assets ?
Myself, taking pain to promote a `compile_command.json` I would, first guess as quick solution, get rid of **dynamic index** being safe then. Any way to do such ? What could be drawback of disabling **dynamic index** : 
- performances issue ?
- kind of editor reactivity loss feeling as end user having file content updated while `compile_command.json` is lagging behind ?
As I could understand `merging symbols seems like a hot codepath` I may foresee **dynamic index** enabling / disabling as a possible "quick & safe" first way to overcome issue. I would say as end user I prefer reliable code indexing against performances.

WDYT ?

**Comment by HighCommander4** [comment]
> So first if my understanding is right we can claim current behavior is related to `clangd` server, not `vscode-clangd` client: do you confirm ?

Yep.

> Considering, yes, some "undesirable" result do you foresee a way to get rid of with current assets ?

Do you mean without changes to the clangd server?

There is a fork of vscode-clangd at https://github.com/eclipse-cdt-cloud/vscode-clangd which has support for using different clangd server instances for different projects (directories) in a vscode workspace.

You may be able to use that to arrange for the `pack-div@1.0.0` directory to use its own clangd instance. Then, opening files in that directory will populate the dynamic index of that clangd instance, and not affect in any way the main clangd instance used for `main.c`.

> Myself, taking pain to promote a `compile_command.json` I would, first guess as quick solution, get rid of **dynamic index** being safe then. Any way to do such ? What could be drawback of disabling **dynamic index** :
> 
>     * performances issue ?
> 
>     * kind of editor reactivity loss feeling as end user having file content updated while `compile_command.json` is lagging behind ?
> 
> 
> As I could understand `merging symbols seems like a hot codepath` I may foresee **dynamic index** enabling / disabling as a possible "quick & safe" first way to overcome issue. I would say as end user I prefer reliable code indexing against performances.
> 
> WDYT ?

Clangd relies fairly heavily on the dynamic index.

To give one notable example, it uses the dynamic index to provide code completion results even for symbols in the current translation unit (more precisely, for symbols defined in the preamble part of the current translation unit (the included headers), so that it doesn't have to do an O(n) traversal of preamble declarations via the AST).

Even if we narrow in on the go-to-definition feature specifically, clangd relies on the dynamic index for workflows such as the following:

 * You are writing a new source/header file pair.
   * The new source file is not reflected in the background index yet, either because you haven't added it to `compile_commands.json` yet, or because clangd hasn't picked up the change to `compile_commands.json` yet (it currently only picks that up on a restart).
 * You perform go-to-def on the declaration of a function you just wrote in the header, and expect it to go to the definition (which you also just wrote in the source file). This relies on the source file being added to the dynamic index.

So, I think that for a solution on the clangd side, rather than disabling the dynamic index altogether, we'll want to change the logic for how information from the dynamic index is combined with information from the background index.

**Comment by VGRSTM** [comment]
Ok @HighCommander4 thanks one time more taking time for explanations.

I'm aware about https://github.com/eclipse-cdt-cloud/vscode-clangd so I' m getting to get an extra try based on. Possibly it may help but to be confirmed (I'll come back to you) because this fork is about `vscode-clangd` not `clangd` server. As we agree on "issue" is part of server ... my fear is despite multi projects capable it may encounter same problematic.
My understanding of https://github.com/eclipse-cdt-cloud/vscode-clangd is to promote capability to support in parallel multiple `compile_commands.json` having each assigned to a `clangd` server instance. I've feeling any server instance may be affected by this **dynamic index** problematic the same.

**Comment by ckandeler** [comment]
It should be noted that the same problem exists even when only one index is involved: If you have a complex project with several applications and/or libraries, then symbols with the same name in the same namespace are indistinguishable (trivial example: main). This is difficult to solve in the general case due to the lack of grouping within compile_commands.json.

**Comment by HighCommander4** [comment]
> I'm aware about https://github.com/eclipse-cdt-cloud/vscode-clangd so I' m getting to get an extra try based on. Possibly it may help but to be confirmed (I'll come back to you) because this fork is about `vscode-clangd` not `clangd` server. As we agree on "issue" is part of server ... my fear is despite multi projects capable it may encounter same problematic. My understanding of https://github.com/eclipse-cdt-cloud/vscode-clangd is to promote capability to support in parallel multiple `compile_commands.json` having each assigned to a `clangd` server instance. I've feeling any server instance may be affected by this **dynamic index** problematic the same.

I think using eclipse-cdt-cloud/vscode-clangd could successfully avoid the problem.

One way to think about the resulting behaviour is this: if you configure `pack-div@1.0.0` to be in a different project than `main.c` (so it uses a different clangd instance), then whenever you open a file in `pack-div@1.0.0`, it will be as if you opened the file in a different vscode window. The behaviour when working with `main.c` will be as if the file in `pack-div@1.0.0` had never been opened.

**Comment by HighCommander4** [comment]
> It should be noted that the same problem exists even when only one index is involved: If you have a complex project with several applications and/or libraries, then symbols with the same name in the same namespace are indistinguishable (trivial example: main). This is difficult to solve in the general case due to the lack of grouping within compile_commands.json.

Yes, the general problem here is indeed a hard one. This is the topic of #1258 (and related issues like #1101).

In this issue, I'm trying to brainstorm ways to mitigate a specific subset of the problem (which was raised by @MNASTM), where only one of the definitions is contained in a file found in `compile_commands.json`, and that's the one that should be preferred over definitions in other files.

**Comment by VGRSTM** [comment]
> > I'm aware about https://github.com/eclipse-cdt-cloud/vscode-clangd so I' m getting to get an extra try based on. Possibly it may help but to be confirmed (I'll come back to you) because this fork is about `vscode-clangd` not `clangd` server. As we agree on "issue" is part of server ... my fear is despite multi projects capable it may encounter same problematic. My understanding of https://github.com/eclipse-cdt-cloud/vscode-clangd is to promote capability to support in parallel multiple `compile_commands.json` having each assigned to a `clangd` server instance. I've feeling any server instance may be affected by this **dynamic index** problematic the same.
> 
> I think using eclipse-cdt-cloud/vscode-clangd could successfully avoid the problem.
> 
> One way to think about the resulting behaviour is this: if you configure `pack-div@1.0.0` to be in a different project than `main.c` (so it uses a different clangd instance), then whenever you open a file in `pack-div@1.0.0`, it will be as if you opened the file in a different vscode window. The behaviour when working with `main.c` will be as if the file in `pack-div@1.0.0` had never been opened.

Yep. Back to be double checked having a try. I agree despite "issue" root source is server contribution (merging indexes), where is maybe not alternative here, acting on client itself (`vscode-clangd` so) having it "filtering" data sent to server can be the way to go.

**Comment by MNASTM** [comment]
Thanks very much for the explanations @HighCommander4.
To summarize we have two tracks to follow here for this "specific subset of the problem" :)

- investigate how to improve the logic for how information from the dynamic index is combined with information from the background index.
- check what scenario can be solved (or not) by the eclipse fork (I'll do that)

**Comment by MNASTM** [comment]
I did some quick tests with the eclipse fork, here is the outcome while trying the multi-projects feature:

- this is partially solving the NOK case (redirection goes to right pack but to .h instead of .c)
- the OK case is now also redirecting to .h instead of .c
- this only work by using multi root workspaces by default

If anyone wants to checkout this [branch ](https://github.com/MNASTM/clangd-issue-682/tree/with-eclipse-clangd-fork)and have a try?

I might be missing something but to me this is still not expected behavior. 
Any chance to investigate on your side the dynamic/background index logic improvement?

**Comment by HighCommander4** [comment]
> I did some quick tests with the eclipse fork, here is the outcome while trying the multi-projects feature:
> 
>    * this is partially solving the NOK case (redirection goes to right pack but to .h instead of .c)
>    * the OK case is now also redirecting to .h instead of .c
>    * this only work by using multi root workspaces by default
> 
> If anyone wants to checkout this [branch ](https://github.com/MNASTM/clangd-issue-682/tree/with-eclipse-clangd-fork)and have a try?
> 
> I might be missing something but to me this is still not expected behavior.

In this branch, both libraries have their own workspace and therefore their own clangd instances. This is why navigation to from the application to a definition in the library does not work.

For it to work in the intended way, you'd want the app + the library version it uses to be in the same workspace (so they share a clangd instance) while the other library version is in a different workspace.

> Any chance to investigate on your side the dynamic/background index logic improvement?

I've added it to my (very long) list of things to look into, but I really can't promise a time frame for that. At the moment there are some folks who have been waiting for months for me to review some of their patches, I need to prioritize that.

**Comment by MNASTM** [comment]
Yes of course, in this case this works, but this way of working is very limiting in terms of package reusability, I think.
Anyway, thanks for your help. Not sure if we can be of any help here, especially if this involved hot codepath you mentioned.
Let me know if the topic comes back to the point anytime.

**Comment by HighCommander4** [comment]
> this way of working is very limiting in terms of package reusability

Could you say more about this?

> Not sure if we can be of any help here

A first step which doesn't require writing any code would be to come up with a precise description of how the heuristic should behave. That will inform questions about implementation and performance.

Note that the example projects in the [main branch](https://github.com/MNASTM/clangd-issue-682/tree/main) and the [eclipse branch](https://github.com/MNASTM/clangd-issue-682/tree/with-eclipse-clangd-fork) differ in an important way:

On the main branch, there is only one `compile_commands.json` file, and of the two library versions, only one of them has its sources included in the `compile_commands.json`. Therefore, only one of the library symbols will be included in clangd's background index.

But on the eclipse branch, there are two applications, both have a `compile_commands.json`, and both include the source files of the corresponding library version. Now, both library sources files are subject to background indexing. If we are working with a single clangd instance, the background index will itself contain a merged symbol, and at merging time we will have chosen one of the two definitions. In this case, the approach discussed above for resolving conflicts between the dynamic index and the background index will not help at all (since the symbol is already merged in the background index). So, if the goal is to get _this_ project setup to work with a single clangd instance (as opposed to the "single compile_commands.json" setup), that will require some new ideas.

**Comment by MNASTM** [comment]
> So, if the goal is to get _this_ project setup to work with a single clangd instance (as opposed to the "single compile_commands.json" setup), that will require some new ideas.

No this was not the goal at all. Here I was just experiencing eclipse fork as suggested to test multi clang instances. And I agree this is actually solving some caveats with multiple `compile_commands.json` at some usage conditions (multiple root workspaces) but this is deviating from the original issue.

> > this way of working is very limiting in terms of package reusability
> 
> Could you say more about this?

I thought it was limiting in a way that you say shared libraries should be in the same workspace, and we don't want to duplicate code just to be part of the workspace, right? But I did another test in this branch but only open both apps (app1 and app2 => I push a new apps.code-workspace). In this case pack-divs are not in the workspace so they won't have their own clang instance. But still I would expect to be redirected. Again, this works, and redirection goes to the right pack-div version but still to the .h file...

Being that said, I think we should not mix issues here and leave the multiple `compile_commands.json` and multiple clangd instances and come back with the original one.

In this single `compile_commands.json` case, my proposed description would be rather simple, and only rephrase the title that you edited just well

"Always prefer definitions from the background index (more reliable), rather than dynamic index."

My approach might be very naïve, and I am not sure how other use cases may be impacted, also probably performance consideration to be studied. WDYT?

**Comment by HighCommander4** [comment]
> > > this way of working is very limiting in terms of package reusability
> > 
> > 
> > Could you say more about this?
> 
> I thought it was limiting in a way that you say shared libraries should be in the same workspace, and we don't want to duplicate code just to be part of the workspace, right? But I did another test in this branch but only open both apps (app1 and app2 => I push a new apps.code-workspace). In this case pack-divs are not in the workspace so they won't have their own clang instance. But still I would expect to be redirected. Again, this works, and redirection goes to the right pack-div version but still to the .h file...
> 
> Being that said, I think we should not mix issues here and leave the multiple `compile_commands.json` and multiple clangd instances and come back with the original one.

Sure. Some final thoughts in this branch of the conversation:

 * I think, in principle, using different clangd instances for different applications could be a way to overcome limitations of the way multiple definitions are handled by a single clangd instance (including limitations that will not be addressed by the heuristic we're discussing below).
 * There's no fundamental reason why the set of files managed by a particular clangd instance would need to align with top-level directory / workspace boundaries. (But the eclipse-cdt-cloud plugin may currently require that.)
 * Getting this to work well in practice would probably require some investigation, and possibly changes to clangd and/or eclipse-cdt-cloud to better accommodate such scenarios.

> In this single `compile_commands.json` case, my proposed description would be rather simple, and only rephrase the title that you edited just well
> 
> "Always prefer definitions from the background index (more reliable), rather than dynamic index."
> 
> My approach might be very naïve, and I am not sure how other use cases may be impacted, also probably performance consideration to be studied. WDYT?

If we always prefer definitions from the background index, we would be discarding the benefit of the dynamic index in providing results which may be more recent than the background index (e.g. a more recent version of the same file, providing a more accurate definition location that reflects recent changes to that file). I don't think we can do that.

I think our heuristic needs to be more narrowly tailored to the case of definitions in files which are not found in the background index.

Let's introduce some terminology:
 * let's call a file which is found in the background index a **project file**
 * let's call a file which is not found in the background index an **external file**

Then a potential heuristic could be:
 * when merging a symbol A from the dynamic index with a symbol B from the background index:
 * if A's definition is in an external file and B's definition is in a project file, then:
    * if the file originating the request (e.g. in the case of go-to-definition, the file in which you invoke the operation) is a project file, prefer B's definition
    * if the file originating the request is an external file, prefer A's definition

This way, go-to-def performed in project files will prefer targets in project files, and go-to-def performed in external files will prefer targets in external files.

Does this behaviour seem reasonable?

**Comment by MNASTM** [comment]
> * There's no fundamental reason why the set of files managed by a particular clangd instance would need to align with top-level directory / workspace boundaries. (But the eclipse-cdt-cloud plugin may currently require that.)

Yes, you are right, see eclipse README:

_A central ProjectService is used to identify each project in an open workspace/folder base on a defined ProjectResolution strategy. Adopters can define custom resolution strategies tailored to their needs (configurable via settings)_

> Does this behaviour seem reasonable?

Sounds good to me yes.

**Comment by VGRSTM** [comment]
@HighCommander4, have you got any chance to look at project / external files ?
Thanks for update if some. 
If helpful, feel free to ask for just in case able to support you a way or another (promoting test case / doing early trial on specific branch / ...). Trying to do the best here, being afraid (pragmatic) unfortunately not being able to contribute to code base itself because looks like real dedicated expertise is must have.

**Comment by HighCommander4** [comment]
I have not, sorry.

I have a day job and I contribute to clangd in my free time. This seems like a relatively easy win, and is one of (several dozen) issues I'd like to work on at some point, but I'm not really in a position to give time estimates.

**Comment by HighCommander4** [comment]
I will add, if someone would like to give this a try before I get to it, I'm happy to provide guidance and code review.

As a starting point, I've already written down a fairly detailed description of the proposed solution at the bottom of [this comment](https://github.com/clangd/clangd/issues/1516#issuecomment-2171185525).

**Comment by TheOnionCutter1** [comment]
I’m working on a related opt-in approach for projects with mutually exclusive implementations across build profiles.
  
 The proposed .clangd option is:
  
```yml
 Index:
    UnlistedSources: Skip
  ```

With Skip, an open source file using an inferred or fallback compile command is still parsed normally, but its symbols are not published to the dynamic index. Files with reliable compile commands and headers remain indexable. Build
preserves the existing behavior and remains the default.

This avoids inactive implementations affecting navigation after switching compile_commands.json. I have a working implementation with unit tests and have verified it in a multi-profile project.

I also saw the contextual symbol-merging approach proposed above: prefer project definitions when navigating from project files, while allowing external files to prefer their own dynamic definitions. Would that approach still be
preferred over an explicit strict option such as UnlistedSources: Skip? I’d be happy to pursue whichever direction fits clangd best.

**Comment by HighCommander4** [comment]
> an open source file using an inferred or fallback compile command is still parsed normally, but its symbols are not published to the dynamic index

Clangd relies heavily on the dynamic index to provide many features for open files. A notable example is that code completion relies on the index to find namespace-scope symbols defined in the preamble portion of an open file (i.e. in included headers), because we don't want to deserialize entire namespaces from the preamble.

You can try this with your implementation. Something like:

foo.h:
```c++
// A file not included by any source file which is background-indexed
void test_function();
```

foo.cpp:
```c++
// A source file which hits the "unlisted" path
#include "foo.h"

void caller() {
  test_f^  // invoke code completion at the ^
}
```

Unless I'm missing something, I would expect that clangd fails to offer `test_function` as a completion item.

---

Due to clangd's reliance on the dynamic index for this and other features, I think that an approach which causes an open file's contents to not be added to the dynamic index at all is too big of a hammer, and would avoid it if other solutions are available.

**Comment by TheOnionCutter1** [comment]
I see, I can reproduce your example with my implementation. Initially, I also had the header open, so it was indexed separately and masked the issue.

I see two possible directions:

1. Keep the UnlistedSources: Skip approach, but suppress only the main-file index contribution while continuing to index the preamble and included headers.
2. Implement the contextual merging approach discussed above: keep all open files fully indexed, but prefer project definitions when navigating from project files and dynamic/external definitions when navigating from external files.

Which direction do you think would fit clangd better?

**Comment by HighCommander4** [comment]
I would lean towards the second because getting the desired behaviour by default means that other users with similar setups benefit from it without having to discover and enable the config option.

I could be convinced otherwise if you try implementing it and find the implementation complexity to be very significant, or find that it doesn't address your use case.

**Comment by TheOnionCutter1** [comment]
I implemented a prototype following your suggestion: clangd keeps the normal merged index globally, but definition lookup uses a merged view that prefers the project-index definition when the dynamic definition comes from a file not
covered by the project index.

This preserves dynamic-index data and header completion while preventing an opened inactive source file from redirecting Go to Definition.

Manual testing works for my multi-configuration project, and the focused clangd tests pass. Does this sound like the right direction for an upstream patch?

**Comment by ArcsinX** [comment]
Solving the problem only for "definition" doesn't seem entirely correct to me. E.g. we will still see "find-all-references" results from **external** files.

It might make sense to add the ability to remove data from the dynamic index. For example, suppose we open an **external** file, then realize we don't need it in the dynamic index and simply close it. Upon `didClose`, we could check if the file is **external** and, if so, remove the index associated with it.

**Comment by TheOnionCutter1** [comment]
That is a good point, though I think removing the file on didClose addresses a slightly different case. It would remove stale dynamic-index data after an external file is closed, but it would not resolve the ambiguity while the file remains open, which is important for multi-configuration projects.

I’m also hesitant to remove all dynamic-index data for external files because they still benefit from diagnostics, completion, document symbols, and header indexing. Whether Find All References should include references from an explicitly opened external file seems less clear to me than definition navigation.
For Go to Definition, the project index provides a natural preference when the definitions conflict.

Perhaps cleanup on didClose would be a useful independent improvement, though this patch keeps the narrower navigation policy.
Do you think Find All References should explicitly exclude results outside the project index as well?

**Comment by ArcsinX** [comment]
> For Go to Definition, the project index provides a natural preference when the definitions conflict.

I’m not sure that kind of logic is natural.
For example, suppose I create a new file and move a definition into it. The new file isn't in the background index, yet the definition within it should be prioritized (which is how it works now).
We are also considering the scenario where the background index has already been fully built, though that might not actually be the case.
The background index could also be outdated.

I generally understand the problem you've described. I’ve often encountered situations where opening a file for a different configuration causes results from that file to appear. I don't need those results, though, and I end up having to restart clangd. My proposal is essentially a substitute for restarting clangd (perhaps as an optional feature that is disabled by default).

> I’m also hesitant to remove all dynamic-index data for external files because they still benefit from diagnostics, completion, document symbols, and header indexing. 

The whole situation seems odd: we don't want certain files to provide definitions, yet we are willing to get other data from them. As I see it, opening an external file is either a deliberate action, such as creating a new file or opening one not yet in the compilation database, in which case we might keep it open. Or it is a temporary action (like checking a file in a different configuration) or perhaps even a mistake (opening a file belonging to a different configuration, meaning it should be closed and its results removed from the index).

> Do you think Find All References should explicitly exclude results outside the project index as well?

I only cited find-all-references as an example. A solution limited to just go-to-definition would seem inconsistent, given that the root "issue" is the priority we assign to the dynamic index. In other words, it affects practically any LSP request.

**Comment by TheOnionCutter1** [comment]
Thanks, those are valid edge cases, particularly when the background index is incomplete or stale.

 I still prefer the contextual merging approach because it best supports the workflow I’m trying to address: inactive implementations may remain open, while unrelated, newly created, or scratch files should continue receiving full
 dynamic-index support.

Removing external files on didClose would be useful, but it only helps after those files are closed. Likewise, suppressing all unlisted sources would incorrectly weaken LSP support for legitimate files outside the compilation database.

I agree that this is ultimately a heuristic: clangd cannot know whether an external file is inactive or newly created. However, when navigation originates from a project file and both indexes provide conflicting definitions, I think
preferring the definition selected by the active compilation database is a reasonable policy. The current behavior is also a heuristic - it always prefers the open file.

I’d therefore still like to pursue this approach, while clearly documenting and testing the stale-background-index trade-off. Of course, I’m open to guidance from the clangd maintainers on whether that trade-off is acceptable upstream.

**Comment by TheOnionCutter1** [comment]
@ArcsinX 
Thanks again for the feedback. I’ve since realized that my previous sentence about being “open to guidance from the clangd maintainers” was poorly worded, since you are one of the maintainers - sorry about that.

For context, the contextual approach came from @HighCommander4's earlier suggestion in this thread, which is why I pursued it. Your stale-index example identifies a trade-off that wasn’t addressed there.

Do you think the current approach has a viable path upstream, perhaps as an explicit opt-in policy for users who consider their compilation database authoritative? Or do you think the navigation-specific policy is fundamentally unsuitable and that I should pursue didClose cleanup instead?

The didClose approach would not fully address my workflow because inactive implementations may remain open. I’d appreciate your guidance on whether the contextual approach can be refined to support that workflow without weakening LSP behavior for ordinary unlisted files.

**Comment by ArcsinX** [comment]
> Do you think the current approach has a viable path upstream, perhaps as an explicit opt-in policy for users who consider their compilation database authoritative? Or do you think the navigation-specific policy is fundamentally unsuitable and that I should pursue didClose cleanup instead?

My main concern relates to inconsistency. With your proposal we are modifying the result only for `go-to-definition`, yet other navigation features use a different approach (including symbol search, which also returns definition of the symbol).
I also took a quick look at your PR, and it makes the code harder to understand. It won't be clear to developers when to use `NavigationIndex` versus `Index`. In short, we are making the code less readable, while the functionality itself remains questionable, there could be cases where the new behavior degrades existing scenarios, such as when the background index becomes stale during operation (though, the dynamic index was created specifically to mitigate this which is why prioritizing the background index feels counterintuitive).

> The didClose approach would not fully address my workflow because inactive implementations may remain open. I’d appreciate your guidance on whether the contextual approach can be refined to support that workflow without weakening LSP behavior for ordinary unlisted files.

I'm simply trying to find a way to solve the originally described problem more generally, without focusing specifically on the `go-to-definition` LSP request (because the problem is not in the definition itself), which is why I am discussing possible options. There may well be other solutions that do not require using `didClose`, I would be happy to discuss them.

**Comment by TheOnionCutter1** [comment]
I think one possible way to avoid preferring potentially stale background data over newer dynamic data is to distinguish dynamic-index contributions from project files and external files.

Requests originating from project files could use the background index together with dynamic data from project files. This would preserve unsaved changes in active sources while preventing open inactive implementations from affecting the active project. Requests originating from external files could additionally use external dynamic data, preserving LSP functionality for scratch and newly created files.

The main implementation question is whether maintaining separate project and external dynamic views, or otherwise filtering the dynamic index by file membership, would fit clangd’s index architecture. Does this seem like a direction worth exploring?

**Comment by ArcsinX** [comment]
> I think one possible way to avoid preferring potentially stale background data over newer dynamic data is to distinguish dynamic-index contributions from project files and external files.

The problem here, that judgment is a file is **project file** or not relies on background index. Even if we have up to date `compile_commands.json` and built background index (which is not always the case), this doesn't mean that we have up to date information about **project file**s, because a header included from **project file** is also **project file**. So, after removing a header inclusion we can face a situation when background index says that this header was indexed, so it's  **project file**, but really it's already not.

I see this issue as a problem where data we don't want there ends up in the dynamic index, and we lack a mechanism to remove it. Specifically, opening a file unrelated to the current project or configuration permanently (until clangd process termination) adds data from that file to the dynamic index.

One more potential solution to consider is replacing existing data with newer data. That is, if we open `A.cpp` containing the definition of function `f()` and subsequently open `B.cpp` containing the definition of `f()`, the definition from `B.cpp` should be the one used. In this way, we can, to some extent, control the presence of unwanted data in the index by overwriting it with the desired data through the opening of the corresponding files.

**Comment by TheOnionCutter1** [comment]
I think I’m framing the desired behavior somewhat differently. My definition of done is:

_Immediately after switching profiles and restarting clangd, Go to Definition should resolve to the implementation selected by the active compilation database, even if source files from inactive profiles remain open._

The solutions you suggested are helpful, but I don’t think they fully satisfy this definition of done.
They require a manual action, such as closing inactive files or reopening the correct implementation.
After restarting clangd, the editor’s buffer-reopening order may also determine which conflicting definition wins.

Do you think there is room to explore another solution that provides this stronger guarantee, or do you consider such profile-aware behavior outside clangd’s scope?

If it is outside the intended scope, I’d still be happy to work on the more general improvements you suggested: removing a file’s dynamic-index contributions on didClose and making newer conflicting contributions replace older ones.

**Comment by ArcsinX** [comment]
> Immediately after switching profiles and restarting clangd, Go to Definition should resolve to the implementation selected by the active compilation database, even if source files from inactive profiles remain open.

Okay, thanks, now I understand your scenario. Architecturally, clangd prioritizes the dynamic index. Therefore, prioritizing the background index in specific cases seems incorrect: making targeted changes to this architecture would complicate code maintenance and require updating the [documentation](https://clangd.llvm.org/design/indexing#implementations) to reflect those specific modifications. Overall, it seems to me that the scenario you described conflicts with a quite common use case: moving a function definition to a new file (where the background index becomes outdated, making it necessary to prioritize the dynamic index, which is exactly how it works now). Ability to modify dynamic index by removing already added data allows to keep original index architecture.

cc @HighCommander4 @timon-ul @JVApen

**Comment by JVApen** [comment]
If I understand the problem correctly, always using the info from the background index is also not OK. Cause if you reindex a file mentioned in the compilation database, you want the more recent info. 

I agree that the use-cases is more exotic and the example of moving a function to a new file makes sense. Though, it is only a problem until the new file is inside the compilation database. Something you most likely want to do anyhow to get it compiling. For me personally, that time window would be rather large as I only start compiling after most red squiggles are gone from my IDE.

To understand the impact on maintenance, I need a mental model of how to solve this. This is what I came up with:

_I haven't checked any feasibility. 
What if we, during parsing, would tag the symbols coming from the source file as being in/out compilation database. This could in the end be a single bit on every symbol. In merge index, you could check the bit. Only if local symbol is out of index and the background isn't, you ignore the local and pass the background as a passthrough._

_The challenge however is when the compilation database changes, as you now have a new list of files and you have to retag all symbols for the new situation. This in every index, including file/remote (also on open)._

_I'd be inclined to say that this could stay rather efficient for the common situations. If you don't want to use the feature, we simply don't tag anything._

_Note: if you move a file and you haven't updated your compilation database, you will keep going to the old non-existing location until you update it and the tagging is completed._

I feel the amount of code that can be contained is rather small, while the impact on the existing code is rather large. I'm mainly thinking about propagating the tagger and some way to run the tagger on every index. Possibly also the merge algorithms in merge index.

Let me remind you of some numbers I've shared already in another issue. We use a file indexes (riff) of 300MB to be loaded. This is a significant large number of symbols. When loading such an index, your delta is the complete compilation database. So it will slow down the loading noticably.

Although I sympathize for the use-case, I'm inclined to say the negative arguments are having more weight. Though at the same time, this doesn't look too complicated to implement either. So, with some time spend, you can implement this on a downstream fork. That would allow you to use this feature even when it doesn't get accepted upstream.

**Comment by TheOnionCutter1** [comment]
Thanks, this mental model matches the behavior I’m trying to achieve.
The current PR is a smaller approximation of this policy: rather than storing a bit on every symbol, it consults the project/background index while resolving conflicting definitions.
However, I understand that this has consistency and stale-index concerns.

A global merge policy based on explicit compilation-database membership seems conceptually cleaner.
The retagging and index-loading costs you described are important, but I wonder whether they can be avoided by tracking membership only for local dynamic-index contributions.
Background, static, and remote indexes are already project-derived, while only dynamic contributions from open files create this conflict.

It may therefore be possible to classify contributions per dynamic file/shard, cache that classification, and rebuild only affected open-file shards when the compilation database changes, rather than tagging or retagging every symbol in every index.
I haven’t verified whether the existing FileIndex preserves enough provenance for this, particularly for headers included through multiple translation units, but I’d be interested in investigating it.

I already use the current PR as a downstream implementation for my personal workflow, so my own problem is solved.
My motivation for contributing it upstream is purely to make this workflow available to other users facing the same problem.

I’ll wait for the others’ opinions before deciding how to proceed.

**Comment by timon-ul** [comment]
After reading these discussion I am reminded of a train of thought I had a while ago while merging a tweak: technically the information if the background index is out of date should have passed through the dynamic index at some point (with reasonable little chance of this not happening). Simple example: we actually want to move `foo()` from `a.c` to a new file  `b.c`. I assume most likely (and please correct me in which way if not) the user will open `a.c` to remove the function and then `b.c` to enter it there. If the first step of opening `a.c` did not happen it seems reasonable to expect that `foo()` sill exists in `a.c` and still is a valid definition.

I am unsure if this is of any help, it might cover that one case where you are moving a definition to a new file, but might not cover other cases. Quite frankly I would need to deepen my understanding for how exactly different cases are handled today and if having 1 bit that tells me if an invalidation has passed through even helps for these cases (and if recording this bit even is easy to do).

Edit: I realize that this has the issue that restarting clangd would lose this information and then return to todays behaviour.

**Comment by TheOnionCutter1** [comment]
@ArcsinX @HighCommander4 

After thinking through it for some time, I think we should acknowledge one constraint: this behavior cannot be achieved purely by the current "dynamic always wins" merge rule.

Also, we wouldn't want to remove/suppress index contributions from open files just because they are not listed in the compilation database.

So for this exact behavior, I think we're left with letting background data win in one narrow case.

First, I want to use a narrower terminology than the one discussed above:
- listed file: a file with a direct command in the current compilation database
- unlisted file: a file without a direct command in the current compilation database

Given a mechanism that tags index contributions by this listed/unlisted distinction:
- dynamic still wins normally.
- background wins only for conflicting definitions where the background definition is from a listed file and the dynamic definition is from an unlisted file.

This is the best I can come up with, I haven't checked how hard this is to implement.

If you think this doesn't match clangd's architecture, I'll happily pursue a more limited solution:
- remove dynamic-index contributions from unlisted files when those files are closed.
- prefer dynamic-index contributions from listed files when conflicting contributions exist.

**Comment by timon-ul** [comment]
Help me out, but doesn't this still break the case where we are creating a new file and facotring out methods from an existing one? Because I do not think we want to break this case.

Edit: I am aware that this is a temporary issue until a new compilation database is generated, but I think a user might feel annoyed by having to do that every time just in favour of an behaviour at least I see as more rare.

**Comment by TheOnionCutter1** [comment]
Thanks, you are correct to point this out.

I think we can address it by refining the rule:
- background wins only for conflicting definitions where:
  - the background definition is from a listed file.
  - the dynamic definition is from an unlisted file.
  - the listed file containing the background definition has not been covered by the dynamic index in the current session/configuration.

This way, if you have `A.c`, create `B.c` and move `foo` from `A.c` to `B.c`, then `A.c` has dynamic coverage.
That dynamic coverage is newer evidence that the old background definition in `A.c` should not be used, so the dynamic definition in `B.c` still wins for Go to Definition.

**Comment by TheOnionCutter1** [comment]
Hi, gentle ping on this when you have a chance.

**Comment by timon-ul** [comment]
I think the problem here is that this approach still invites a lot of problems into the code base, we will have to know where the Symbols come from and this most likely needs a lot more storage used up (or slower). Also the logic of the index gets more complicated. I think from clangd side we would prefer a solution that is rather lightweight and simple optimally, otherwise we need to consider if this is really worth adding (and I think so far the opinion seems rather negative) or at best it will be opt-in (if the code changes are not too invasive).

I have to admit though, I am unsure if the idea by @ArcsinX to clean up after closing a file is so simple either, you still have to be aware which symbols have to be removed later on, so this information has to be stored somewhere and at least to me this seems like it might already cause some annoying code changes (unless I am missing a simple way).

One more idea I came up is, what if instead of worrying about removing data or when to use what available data from which index, we try harder in preventing unwanted information getting into the dynamic index in the first place? I am unsure how easy to achieve this is, but at least to me this seems like the most desireable way. It would also mean we do not need to worry about where the Symbols live, so no additional data saved in the index. I think it follows a similar way like the removal idea in where we can keep the overall architecture the same. The bigger problem though might be how you determine what symbols are allowed into the index, at least to me it seems like it might be tough to differenciate between the case of just creating a new file with a new definition and opening a file with a definition that already exists in the background index.

No matter the approach, I think it is hard to judge any of them just based on a rough idea without anyone spending the time to prototype them tough.
